src/Nethermind/Nethermind.Synchronization/SnapSync/SnapProvider.cs

```diff
@@ -0,0 +1,294 @@
+using Nethermind.Db;
+using Nethermind.Logging;
+using Nethermind.State;
+using Nethermind.State.Snap;
+using Nethermind.Trie;
+using Nethermind.Trie.Pruning;
+
+namespace Nethermind.Synchronization.SnapSync
+{
+    public class SnapProvider : ISnapProvider
+    {
+        private readonly ITrieStore _store;
+        private readonly IDbProvider _dbProvider;
+        private readonly ILogManager _logManager;
+        private readonly ILogger _logger;
+
+        private readonly ProgressTracker _progressTracker;
+
+        public SnapProvider(ProgressTracker progressTracker, IDbProvider dbProvider, ILogManager logManager)
+        {
+            _dbProvider = dbProvider ?? throw new ArgumentNullException(nameof(dbProvider));
+            _progressTracker = progressTracker ?? throw new ArgumentNullException(nameof(progressTracker));
+
+            _store = new TrieStore(
+                _dbProvider.StateDb,
+                No.Pruning,
+                Persist.EveryBlock,
+                logManager);
+
+            _logManager = logManager ?? throw new ArgumentNullException(nameof(logManager));
+            _logger = logManager.GetClassLogger();
+        }
+
+        public bool CanSync() => _progressTracker.CanSync();
+
+        public (SnapSyncBatch request, bool finished) GetNextRequest() => _progressTracker.GetNextRequest();
+
+        public AddRangeResult AddAccountRange(AccountRange request, AccountsAndProofs response)
+        {
+            AddRangeResult result;
+
+            if (response.PathAndAccounts.Length == 0 && response.Proofs.Length == 0)
+            {
+                _logger.Trace($"SNAP - GetAccountRange - requested expired RootHash:{request.RootHash}");
+
+                result = AddRangeResult.ExpiredRootHash;
+            }
+            else
+            {
+                result = AddAccountRange(request.BlockNumber.Value, request.RootHash, request.StartingHash, response.PathAndAccounts, response.Proofs);
+
+                if (result == AddRangeResult.OK)
+                {
+                    Interlocked.Add(ref Metrics.SnapSyncedAccounts, response.PathAndAccounts.Length);
+                }
+            }
+
+            _progressTracker.ReportAccountRequestFinished();
+
+            return result;
+        }
+
+        public AddRangeResult AddAccountRange(long blockNumber, Keccak expectedRootHash, Keccak startingHash, PathWithAccount[] accounts, byte[][] proofs = null)
+        {
+            StateTree tree = new(_store, _logManager);
+
+            (AddRangeResult result, bool moreChildrenToRight, IList<PathWithAccount> accountsWithStorage, IList<Keccak> codeHashes) =
+                SnapProviderHelper.AddAccountRange(tree, blockNumber, expectedRootHash, startingHash, accounts, proofs);
+
+            if (result == AddRangeResult.OK)
+            {
+                foreach (var item in accountsWithStorage)
+                {
+                    _progressTracker.EnqueueAccountStorage(item);
```

Comment:
we are allocating accountsWithStorage just to iterate them and add to next collection